### PR TITLE
HPCC-16637 Save old environment when update detected.

### PR DESCRIPTION
**Pull request by jakesmith**
Keep a copy of the old environment, to help diagnosing when
issues/changes are related to environment reconfiguration.

Signed-off-by: Jake Smith <jake.smith@lexisnexis.com>

### CODE REVIEW

**Comment by jakesmith** [comment]
@richardkchapman - please review

File: dali/base/dasds.cpp [hunk 1]
```diff
@@ -5947,6 +5948,18 @@ static int extNcompareFunc(CInterface * const *_itm1, CInterface * const *_itm2)
...
 
+void CCovenSDSManager::checkEnvironment(IPropertyTree *oldEnv, IPropertyTree *newEnv)
+{
+    if (!areMatchingPTrees(oldEnv, newEnv))
+    {
+        StringBuffer filename("environment");
+        addFileTimestamp(filename);
+        filename.append(".xml");
+        PROGLOG("Detected environment change, backing up previous version to: %s", filename.str());
+        saveXML(filename, oldEnv, 0, XML_Format|XML_SortTags); // sort for easier diffing
```
Comment (richardkchapman) [inline dali/base/dasds.cpp]: I had envisaged the old environments would be saved within Dali rather than as separate files on disk.

**Comment by richardkchapman** [comment]
@jakesmith One comment

**Comment by jakesmith** [comment]
That would be rather expensive to Dali over time.. though we could keep N where N is smallish I suppose.

You are worried that if on disk, ops. might clean away?
Storing them in the Dali data directory would be better... this PR currently saves them in the instance directory.
If in the Dali data directory I don't think anyone will touch them.

@richardkchapman - what do you think?

**Comment by richardkchapman** [comment]
If they are larger than some threshold then they will be stored on disk anyway even if they appear within the dali tree, correct? I don't know what that threshold is though. If it's smaller than the typical environment file then there's really not a lot of difference between storing in dali vs storing in the dali directory as far as growth over time is concerned. How often does the environment change, anyway?

I think I'd vote for storing in dali and having sasha clean away old ones if we feel that is necessary.

@jakesmith

**Comment by jakesmith** [comment]
>If they are larger than some threshold then they will be stored on disk anyway even if they appear within the dali tree, correct?

That's only true if a single XML value exceeds the threshold (which is 4k by default).
But yes, I could store old copies of the environment like that, then they work they will be serialized to disk.
In which case it wouldn't have much impact on Dali's memory footprint.
Dali (not Sasha) which is saving away the old versions, can have a setting to limit the number kept.

I'll re-implement like that.

**Comment by jakesmith** [comment]
@richardkchapman - as discussed, updated so that it saves copies of old environments in Dali's data directory, as data stamped XML files.

Please review/merge.

**Pull request merged**